sysgen/sysgen.go

```diff
@@ -249,66 +249,97 @@ func generateResources(desc *Description, consts map[string]uint64, out io.Write
 	fmt.Fprintf(out, "}\n")
 }
 
-func generateStructs(desc *Description, consts map[string]uint64, out io.Writer) {
-	var structArray StructArray
-	for _, str := range desc.Structs {
-		structArray = append(structArray, str)
+func generateStructEntry(str Struct, key string, name string, out io.Writer) {
+	typ := "StructType"
+	if str.IsUnion {
+		typ = "UnionType"
+	}
+	packed := ""
+	if str.Packed {
+		packed = ", packed: true"
+	}
+	varlen := ""
+	if str.Varlen {
+		varlen = ", varlen: true"
+	}
+	align := ""
+	if str.Align != 0 {
+		align = fmt.Sprintf(", align: %v", str.Align)
+	}
+	fmt.Fprintf(out, "\"%v\": &%v{TypeCommon: TypeCommon{TypeName: \"%v\", IsOptional: %v} %v %v %v},\n",
+		key, typ, name, false, packed, align, varlen)
+}
+
+func generateStructFields(str Struct, ref string, desc *Description, consts map[string]uint64, out io.Writer) {
+	typ := "StructType"
+	fields := "Fields"
+	if str.IsUnion {
+		typ = "UnionType"
+		fields = "Options"
+	}
+	fmt.Fprintf(out, "{ s := %v.(*%v)\n", ref, typ)
+	for _, a := range str.Flds {
+		fmt.Fprintf(out, "s.%v = append(s.%v, ", fields, fields)
+		generateArg(str.Name, a[0], a[1], a[2:], desc, consts, false, true, out)
+		fmt.Fprintf(out, ")\n")
 	}
-	sort.Sort(structArray)
+	fmt.Fprintf(out, "}\n")
+}
 
+func generateStructs(desc *Description, consts map[string]uint64, out io.Writer) {
 	// Struct fields can refer to other structs. Go compiler won't like if
 	// we refer to Structs map during Structs map initialization. So we do
 	// it in 2 passes: on the first pass create types and assign them to
 	// the map, on the second pass fill in fields.
-	fmt.Fprintf(out, "var Structs = map[string]Type{\n")
-	for _, str := range structArray {
-		typ := "StructType"
-		if str.IsUnion {
-			typ = "UnionType"
-		}
-		packed := ""
-		if str.Packed {
-			packed = ", packed: true"
-		}
-		varlen := ""
-		if str.Varlen {
-			varlen = ", varlen: true"
+
+	// Since structs of the same type can be fields with different names
+	// of multiple other structs, we have an instance of those structs
+	// for each field indexed by the name of the parent struct and the
+	// field name.
+
+	structMap := make(map[string]Struct)
+	for _, str := range desc.Structs {
+		if _, ok := structMap[str.Name]; ok {
+			failf("two structs with the same name '%v'\n", str.Name)
 		}
-		align := ""
-		if str.Align != 0 {
-			align = fmt.Sprintf(", align: %v", str.Align)
+		structMap[str.Name] = str
```

Comment:
I think it will be better if you add all structs to structMap here:

```go
	for _, a := range str.Flds {
			if innerStr, ok := desc.Structs[a[1]]; ok {
				structMap[fmt.Sprintf("%v-%v", str.Name, a[0])] = str
			}
  	}
```

then you don't need duplicate loops below. You can just go over structMap.